File: matiasmasso/Mat.NET/DTO/DTOCsb.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOCsb : DTOBaseGuid
    {
        public DTOCsa Csa { get; set; }
        public int Id { get; set; }
        public DTOContact Contact { get; set; }
        public string Txt { get; set; }
        public DTOAmt Amt { get; set; }
        public DateTime Vto { get; set; }
        public DTOIban Iban { get; set; }
        public string FraNum { get; set; }
        public int FraYea { get; set; }

        // Property Reclamat As Boolean
        // Property Impagat As Boolean

        public Results Result { get; set; }
        public DTOCca ResultCca { get; set; }

        public DTOInvoice Invoice { get; set; }
        public DTOPnd Pnd { get; set; }
        public TiposAdeudo SepaTipoAdeudo { get; set; }

        public ExceptionCodes ExceptionCode { get; set; }

        public DTOAmt Despesa { get; set; }

        public List<DTOMailingLog> MailingLogs { get; set; }

        public enum TiposAdeudo
        {
            NotSet,
            FRST,
            RCUR,
            FNAL,
            OOFF
        }

        public enum ExceptionCodes
        {
            Success,
            NoValue,
            Negative,
            NoIban,
            NoBn2,
            NoBn1,
            WrongBIC,
            NoMandate,
            NoMandateFch
        }

        public enum Results
        {
            Pendent,
            Vençut,
            Reclamat,
            Impagat
        }

        public DTOCsb() : base()
        {
            this.MailingLogs = new List<DTOMailingLog>();
        }

        public DTOCsb(Guid oGuid) : base(oGuid)
        {
            this.MailingLogs = new List<DTOMailingLog>();
        }

        public string FormattedId()
        {
            string retval = string.Format("{0}{1:000}", this.Csa.formattedId(), this.Id);
            return retval;
        }

        public string ReadableFormat()
        {
            string retval = string.Format("{0}.{1}", this.Csa.readableFormat(), this.Id);
            return retval;
        }

        public static DTOCountry Country(DTOCsb oCsb)
        {
            DTOCountry retval = null/* TODO Change to default(_) if this is not a reference type */;
            if (oCsb != null)
            {
                if (oCsb.Iban != null)
                {
                    if (oCsb.Iban.BankBranch != null)
                    {
                        if (oCsb.Iban.BankBranch.Bank != null)
                            retval = oCsb.Iban.BankBranch.Bank.Country;
                    }
                }
            }
            return retval;
        }

        public static bool validate(List<DTOCsb> oCsbs, DTOCsa.FileFormats oFormat)
        {
            bool retval = true;
            foreach (DTOCsb item in oCsbs)
            {
                if (!DTOCsb.validate(item, oFormat))
                    retval = false;
            }
            return retval;
        }

        public static bool validate(DTOCsb oCsb, DTOCsa.FileFormats oFormat)
        {
            // If oCsb.Iban.Digits.StartsWith("AD55") Then Stop
            {
                var withBlock = oCsb;
                withBlock.ExceptionCode = DTOCsb.ExceptionCodes.Success;
                if (withBlock.Amt.Eur == 0)
                    withBlock.ExceptionCode = DTOCsb.ExceptionCodes.NoValue;
                else if (withBlock.Amt.Eur < 0)
                    withBlock.ExceptionCode = DTOCsb.ExceptionCodes.Negative;
                else if (withBlock.Iban == null)
                    withBlock.ExceptionCode = DTOCsb.ExceptionCodes.NoIban;
                else if (withBlock.Iban.BankBranch == null)
                    withBlock.ExceptionCode = DTOCsb.ExceptionCodes.NoBn1;
                else if (withBlock.Iban.BankBranch.Bank == null)
                    withBlock.ExceptionCode = DTOCsb.ExceptionCodes.NoBn2;
                else if (!DTOBank.validateBIC(withBlock.Iban.BankBranch.Bank.Swift))
                    withBlock.ExceptionCode = DTOCsb.ExceptionCodes.WrongBIC;
                else if (oFormat != DTOCsa.FileFormats.normaAndorrana & oFormat != DTOCsa.FileFormats.remesesExportacioLaCaixa & withBlock.Iban.FchFrom == null/* TODO Change to default(_) if this is not a reference type */ )
                    withBlock.ExceptionCode = DTOCsb.ExceptionCodes.NoMandateFch;
                else if (oFormat != DTOCsa.FileFormats.normaAndorrana & oFormat != DTOCsa.FileFormats.remesesExportacioLaCaixa & withBlock.Iban.DocFile == null)
                    withBlock.ExceptionCode = DTOCsb.ExceptionCodes.NoMandate;
            }
            bool retval = oCsb.ExceptionCode == DTOCsb.ExceptionCodes.Success;
            return retval;
        }

        public static string ValidationText(DTOCsb.ExceptionCodes oCode)
        {
            string retval = "";
            switch (oCode)
            {
                case DTOCsb.ExceptionCodes.NoValue:
                    {
                        retval = "sense import";
                        break;
                    }

                case DTOCsb.ExceptionCodes.Negative:
                    {
                        retval = "import negatiu";
                        break;
                    }

                case DTOCsb.ExceptionCodes.NoIban:
                    {
                        retval = "sense Iban vigent";
                        break;
                    }

                case DTOCsb.ExceptionCodes.NoBn1:
                    {
                        retval = "entitat bancaria no registrada";
                        break;
                    }

                case DTOCsb.ExceptionCodes.NoBn2:
                    {
                        retval = "oficina bancaria no registrada";
                        break;
                    }

                case DTOCsb.ExceptionCodes.WrongBIC:
                    {
                        retval = "BIC incorrecte a entitat bancaria";
                        break;
                    }

                case DTOCsb.ExceptionCodes.NoMandateFch:
                    {
                        retval = "mandat sense data";
                        break;
                    }

                case DTOCsb.ExceptionCodes.NoMandate:
                    {
                        retval = "sense document de mandat";
                        break;
                    }
            }
            return retval;
        }


        public static DTOAmt TotalNominal(List<DTOCsb> oCsbs)
        {
            decimal DcEur = oCsbs.Sum(x => x.Amt.Eur);
            DTOAmt retval = DTOAmt.Factory(DcEur);
            return retval;
        }

        public static DTOAmt TotalDespeses(List<DTOCsb> oCsbs, DTOBancTerm oTerm)
        {
            var retval = DTOAmt.Empty();
            foreach (DTOCsb oCsb in oCsbs)
                retval.Add(DTOBancTerm.cost(oCsb, oTerm));
            return retval;
        }
    }
}

File: matiasmasso/Mat.NET/DT2/DTOCsb.vb
Public Class DTOCsb
    Inherits DTOBaseGuid

    Property csa As DTOCsa
    Property id As Integer
    Property contact As DTOContact
    Property txt As String
    Property amt As DTOAmt
    Property vto As Date
    Property iban As DTOIban
    Property fraNum As String
    Property fraYea As Integer

    'Property Reclamat As Boolean
    'Property Impagat As Boolean

    Property result As Results
    Property resultCca As DTOCca

    Property invoice As DTOInvoice
    Property pnd As DTOPnd
    Property sepaTipoAdeudo As TiposAdeudo

    Property exceptionCode As ExceptionCodes

    Property despesa As DTOAmt

    Property mailingLogs As List(Of DTOMailingLog)

    Public Enum TiposAdeudo
        NotSet
        FRST
        RCUR
        FNAL
        OOFF
    End Enum

    Public Enum ExceptionCodes
        Success
        NoValue
        Negative
        NoIban
        NoBn2
        NoBn1
        WrongBIC
        NoMandate
        NoMandateFch
    End Enum

    Public Enum Results
        Pendent
        Vençut
        Reclamat
        Impagat
    End Enum

    Public Sub New()
        MyBase.New()
        _MailingLogs = New List(Of DTOMailingLog)
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
        _MailingLogs = New List(Of DTOMailingLog)
    End Sub

    Public Function FormattedId() As String
        Dim retval As String = String.Format("{0}{1:000}", _Csa.formattedId, _Id)
        Return retval
    End Function

    Public Function ReadableFormat() As String
        Dim retval As String = String.Format("{0}.{1}", _Csa.ReadableFormat, _Id)
        Return retval
    End Function

    Shared Function Country(oCsb As DTOCsb) As DTOCountry
        Dim retval As DTOCountry = Nothing
        If oCsb IsNot Nothing Then
            If oCsb.Iban IsNot Nothing Then
                If oCsb.Iban.BankBranch IsNot Nothing Then
                    If oCsb.Iban.BankBranch.Bank IsNot Nothing Then
                        retval = oCsb.Iban.BankBranch.Bank.Country
                    End If
                End If
            End If
        End If
        Return retval
    End Function

    Shared Function Validate(oCsbs As List(Of DTOCsb), oFormat As DTOCsa.FileFormats) As Boolean
        Dim retval As Boolean = True
        For Each item As DTOCsb In oCsbs
            If Not DTOCsb.Validate(item, oFormat) Then retval = False
        Next
        Return retval
    End Function

    Shared Function Validate(ByRef oCsb As DTOCsb, oFormat As DTOCsa.FileFormats) As Boolean
        'If oCsb.Iban.Digits.StartsWith("AD55") Then Stop
        With oCsb
            .ExceptionCode = DTOCsb.ExceptionCodes.Success
            If .Amt.Eur = 0 Then
                .ExceptionCode = DTOCsb.ExceptionCodes.NoValue
            ElseIf .Amt.Eur < 0 Then
                .ExceptionCode = DTOCsb.ExceptionCodes.Negative
            ElseIf .Iban Is Nothing Then
                .ExceptionCode = DTOCsb.ExceptionCodes.NoIban
            ElseIf .Iban.BankBranch Is Nothing Then
                .ExceptionCode = DTOCsb.ExceptionCodes.NoBn1
            ElseIf .Iban.BankBranch.Bank Is Nothing Then
                .ExceptionCode = DTOCsb.ExceptionCodes.NoBn2
            ElseIf Not DTOBank.ValidateBIC(.Iban.BankBranch.Bank.Swift) Then
                .ExceptionCode = DTOCsb.ExceptionCodes.WrongBIC
            ElseIf oFormat <> DTOCsa.FileFormats.NormaAndorrana And oFormat <> DTOCsa.FileFormats.RemesesExportacioLaCaixa And .Iban.FchFrom = Nothing Then
                .ExceptionCode = DTOCsb.ExceptionCodes.NoMandateFch
            ElseIf oFormat <> DTOCsa.FileFormats.NormaAndorrana And oFormat <> DTOCsa.FileFormats.RemesesExportacioLaCaixa And .Iban.DocFile Is Nothing Then
                .ExceptionCode = DTOCsb.ExceptionCodes.NoMandate
            End If
        End With
        Dim retval As Boolean = oCsb.ExceptionCode = DTOCsb.ExceptionCodes.Success
        Return retval
    End Function

    Shared Function ValidationText(oCode As DTOCsb.ExceptionCodes) As String
        Dim retval As String = ""
        Select Case oCode
            Case DTOCsb.ExceptionCodes.NoValue
                retval = "sense import"
            Case DTOCsb.ExceptionCodes.Negative
                retval = "import negatiu"
            Case DTOCsb.ExceptionCodes.NoIban
                retval = "sense Iban vigent"
            Case DTOCsb.ExceptionCodes.NoBn1
                retval = "entitat bancaria no registrada"
            Case DTOCsb.ExceptionCodes.NoBn2
                retval = "oficina bancaria no registrada"
            Case DTOCsb.ExceptionCodes.WrongBIC
                retval = "BIC incorrecte a entitat bancaria"
            Case DTOCsb.ExceptionCodes.NoMandateFch
                retval = "mandat sense data"
            Case DTOCsb.ExceptionCodes.NoMandate
                retval = "sense document de mandat"
        End Select
        Return retval
    End Function


    Shared Function TotalNominal(oCsbs As List(Of DTOCsb)) As DTOAmt
        Dim DcEur As Decimal = oCsbs.Sum(Function(x) x.Amt.Eur)
        Dim retval As DTOAmt = DTOAmt.factory(DcEur)
        Return retval
    End Function

    Shared Function TotalDespeses(oCsbs As List(Of DTOCsb), oTerm As DTOBancTerm) As DTOAmt
        Dim retval = DTOAmt.Empty
        For Each oCsb As DTOCsb In oCsbs
            retval.Add(DTOBancTerm.Cost(oCsb, oTerm))
        Next
        Return retval
    End Function

End Class
